service/frontend/workflow_handler.go

```diff
@@ -3229,15 +3214,40 @@ func (wh *WorkflowHandler) DescribeSchedule(ctx context.Context, request *workfl
 		return errWaitForRefresh
 	}
 
-	// TODO: confirm retry is necessary here.
-	policy := backoff.NewExponentialRetryPolicy(50 * time.Millisecond)
+	// wait up to 4 seconds or rpc deadline minus 1 second, but at least 1 second
+	expiration := 4 * time.Second
+	if deadline, ok := ctx.Deadline(); ok {
+		remaining := deadline.Sub(time.Now()) - 1*time.Second
+		expiration = util.Min(expiration, remaining)
+	}
+	expiration = util.Max(expiration, 1*time.Second)
+	policy := backoff.NewExponentialRetryPolicy(200 * time.Millisecond).
+		WithExpirationInterval(expiration)
 	isWaitErr := func(e error) bool { return e == errWaitForRefresh }
+
 	err = backoff.ThrottleRetryContext(ctx, op, policy, isWaitErr)
-	if err != nil {
+	// if we still got errWaitForRefresh that means we used up our retries, just return
+	// whatever we have
```

Comment:
I think errWaitForRefresh should be retried only once. Or the response should have some last update timestamp and if it is less than some threshold, we should not keep retrying.

Reply:
It's a little confusing because it's retrying for two different reasons: 1. the workflow hasn't received or acted on the signal yet, 2. it has acted but the query result lists _new_ running workflows that we think are not running.

In case 2, yes, we could just not send another signal and return what we have. But in case 1 we should keep retrying some more.

We don't need a timestamp, we can detect the cases based on a change in the returned set of workflows, i.e. if we see a new one that shouldn't be there.

Reply:
We can also be a little sneaky and instead of retrying the query blindly, do a long-poll PollMutableState to wait for more events to appear. That could be more efficient